OsmAnd/src/net/osmand/plus/views/mapwidgets/widgets/AlarmWidget.java

```diff
@@ -138,29 +143,60 @@ public boolean updateInfo(DrawSettings drawSettings, boolean drawBitmap) {
 							icon.setImageResource(info.locImgId);
 						}
 					}
-					if (!Algorithms.objectEquals(info.text, cachedText) || cachedRegion != info.region) {
+
+					boolean isRegionChanged = cachedRegion != info.region;
+					if (isRegionChanged) {
+						cachedRegion = info.region;
+					}
+
+					boolean isAlarmTypeChanged = cachedAlarmType != alarm.getType();
+					if (isAlarmTypeChanged) {
+						cachedAlarmType = alarm.getType();
+					}
+
+					if (!Algorithms.objectEquals(info.text, cachedText) || isRegionChanged || isAlarmTypeChanged) {
 						changed = true;
 						cachedText = info.text;
-						cachedRegion = info.region;
 						if (layout != null && widgetText != null) {
 							widgetText.setText(cachedText);
 							Resources res = layout.getContext().getResources();
-							if (alarm.getType() == AlarmInfo.AlarmInfoType.SPEED_LIMIT && info.americanType && !info.isCanadianRegion) {
+							if (cachedAlarmType == AlarmInfo.AlarmInfoType.SPEED_LIMIT && info.americanType && !info.isCanadianRegion) {
 								int topPadding = res.getDimensionPixelSize(R.dimen.map_alarm_text_top_padding);
 								widgetText.setPadding(0, topPadding, 0, 0);
 							} else {
 								widgetText.setPadding(0, 0, 0, 0);
 							}
+
```

Comment:
this is already quite big chunk of code so, it makes sense I think to move if block inside to a separate method